lib/puppet/environments.rb

```diff
@@ -230,20 +233,22 @@ def create_environment(name)
       env
     end
 
-    def valid_directory?(envdir)
-      name = Puppet::FileSystem.basename_string(envdir)
-      Puppet::FileSystem.directory?(envdir) &&
-         Puppet::Node::Environment.valid_name?(name)
+    def validated_directory(envdir)
+      env_name = Puppet::FileSystem.basename_string(envdir)
+      envdir = Puppet::Environments::Directories.real_path(envdir)
+      return nil unless Puppet::FileSystem.directory?(envdir) && Puppet::Node::Environment.valid_name?(env_name)
+      return envdir
     end
 
-    def valid_directories
+    def valid_environments
```

Comment:
why the name change?

Reply:
eh, I had refactored this a bunch and `valid_directories` didn't seem as accurate anymore, though now I see that I landed more closely to the original than I had in some in-between iterations. We're returning a list of environment names, so this seemed clearer to me, but I can put it back if you'd prefer.

Reply:
I expected this to be a list of `Environment`s, not to be super pedantic but I think we should name things to keep "whole paths", "directory names", and actual Environments separate.

Reply:
okay, that's fair. I guess when I think of a "directory" I imagine a path? What do you think of `valid_environment_directories`? or `valid_environment_directory_names`? Are those too wordy? Maybe I'll just put it back to how it was..?

Reply:
I don't think they're too wordy, `valid_environment_names` would work too?

Reply:
I like `valid_environment_names` 😊

Reply:
updated